src/mame/machine/k573mcr.cpp

```diff
@@ -1,24 +1,466 @@
+  |-------------------------------|
 
-/*
-  GE885-PWB(A)A ( contains Toshiba tmpr3904af, ram, rom, tranceiver and glue ).
+Notes:
+	DIP8       - 8-position DIP switch
+	CN61       - BS8PSHF1AA 8 pin connector, connects to memory card harness
+	CN62       - BS8PSHF1AA 8 pin connector
+	CN63       - 6P-SHVQ labeled "0", GE885-JB security dongle is connected here
+	CN64       - 6P-SHVQ labeled "1"
+	CN65       - B4PS-VH, 4 pin power connector
+	CN67       - BS15PSHF1AA, 15-pin connector, unpopulated
+	USB-A      - USB-A connector
+	USB-B      - USB-B connector, connects to USB on System 573 motherboard
+	ADM485JR   - Analog Devices ADM485 low power EIA RS-485 transceiver
+	TMPR3904AF - Toshiba TMPR3904AF RISC Microprocessor
+	XCS05/10   - XILINX XCS10XL VQ100AKP9909 A2026631A
+	DRAM4M     - Silicon Magic 66 MHz C9742 SM81C256K16CJ-35, 256K x 16 EDO DRAM
+	EP4M16     - ROM labeled "855-A01"
 */
 
+#include "emu.h"
+#include "k573mcr.h"
+
 k573mcr_device::k573mcr_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
-	device_t(mconfig, KONAMI_573_MEMORY_CARD_READER, tag, owner, clock)
+	jvs_device(mconfig, KONAMI_573_MEMORY_CARD_READER, tag, owner, clock),
+	m_ports{{*this, "port1"}, {*this, "port2"}}
 {
+	m_ram = std::make_unique<uint8_t[]>(RAM_SIZE);
 }
 
 void k573mcr_device::device_start()
 {
+	jvs_device::device_start();
+
+	save_pointer(NAME(m_ram), RAM_SIZE);
+	save_item(NAME(m_is_memcard_initialized));
+	save_item(NAME(m_status));
+}
+
+void k573mcr_device::device_reset()
+{
+	jvs_device::device_reset();
+
+	memset(m_ram.get(), 0, RAM_SIZE);
+	m_is_memcard_initialized = false;
+	m_status = 0;
+}
+
+void k573mcr_device::device_add_mconfig(machine_config &config)
+{
+	// The actual controllers are only used in Guitar Freaks and it was
+	// meant to be used with the home version PS1 guitar controller.
+	// The PS1 guitar controller isn't emulated in MAME and it doesn't
+	// make much sense to have it enabled by default. You can still select
+	// a controller through the slots menu.
+	// The memory card ports are still usable even without a controller
+	// enabled which is the main reason for using the PSX controller ports.
+	PSX_CONTROLLER_PORT(config, m_ports[0], psx_controllers, nullptr);
+	PSX_CONTROLLER_PORT(config, m_ports[1], psx_controllers, nullptr);
+}
+
+const char *k573mcr_device::device_id()
+{
+	return "KONAMI CO.,LTD.;White I/O;Ver1.0;White I/O PCB";
+}
+
+uint8_t k573mcr_device::command_format_version()
+{
+	return 0x11;
+}
+
+uint8_t k573mcr_device::jvs_standard_version()
+{
+	return 0x20;
+}
+
+uint8_t k573mcr_device::comm_method_version()
+{
+	return 0x10;
+}
+
+uint8_t k573mcr_device::controller_port_send_byte(uint32_t port_no, uint8_t data)
+{
+	auto port = m_ports[port_no];
+	uint8_t output = 0;
+
+	for (int i = 0; i < 8; i++) {
+		port->clock_w(0);
+		port->tx_w(!!(data & (1 << i)));
+		port->clock_w(1);
+		output |= port->rx_r() << i;
+	}
+
+	return output;
+}
```

Comment:
Is there a better way to implement communication with the PS1 controller ports? I didn't think it was appropriate to reimplement something separate entirely because the hardware uses real PSX controller ports. My other idea was to expose `psxcard_device::transfer` and `psx_standard_controller_device::get_pad` but I don't feel comfortable making that change on my own.

Reply:
Well it actually *is* serial, and transfers take time.  Pretending it happens instantaneously like this is bad enough; pretending it has a parallel interface would be even worse.

Reply:
I rewrote the PSX code to use the PSX controller bus instead of communicating to the ports directly in preparation for trying to implement the PSX code in a more proper way that took transfer time into consideration. The memory card part worked fine because the game polls waiting for a status flag and will just keep polling until the read/write operation has finished, but I ran into a big problem when I tried to implement the pad reads. The pad read function expects to be able to return the pad status as the response to the pad read command. It does not rely on polling the status flag like memory card functions. I think a more skilled person would be able to implement this part in a way that respects transfer times, but that level of implementation is currently outside of my skill level.

Reply:
Don't let yourself get intimidated by the codebase. Given the level of skill you've shown in turning around nearly all of the feedback points, I think you're more than skilled enough to handle it.

As part of its core, MAME has a class called emu_timer. It's not meant to encapsulate any particular timer in a chip, it's just intended for use when you want to fire a callback a given amount of time in the future. It can either be implemented using a device's built-in `device_timer` callback (preferable if a device only has one thing that needs to be timed), or using a specifically-named `TIMER_CALLBACK_MEMBER` function, which is preferable in the case of a device that has multiple potential timing-related tasks that need to be coordinated.

If you need an example on how to make use of it, I would refer to the SA1110 or SA1111 devices, both of which make heavy use of delayed operations.

If you have any questions, we're here to help.

Reply:
@MooglyGuy Thanks for the comment. I will check out the devices you mentioned. I actually was using `TIMER_CALLBACK_MEMBER` in my attempt. I pushed the WIP code to a separate branch if you're willing to look at it and give me feedback on what I could do to fix it up. Everything except the pad reporting functionality is working, and reading the pad state itself actually works even. I just can't figure out how to get the data out of the buffer fast enough to return in the response for `handle_message`.

https://github.com/987123879113/mame/commit/5053cb099e808242a4fc9b9865c145bbcface81b

This is kind of a slow realization that I made while rereading my code for this response, but I just realized that I don't need to poll the pad when it's requested. I can just poll for pad status updates in the timer update and store the results for when it's requested and that should probably solve the issue for the most part.

Reply:
Your realization sounds about right. In reality, the actual pad status is most likely only latched when the status changes, and the timer should only be kicked off when a new request comes in. If there's a request currently in-flight, then presumably the status register would indicate that, without the need for kicking off a new read cycle.